### ISSUE TITLE
Error in InsertionBias.ChromatinAssay

### ISSUE THREAD

**Issue by Jackieshen** [timoast/signac#964]
Dear Timoast and Signac Team,

I've tried to run Footprint for different TFs several times, but got identical error messages shown below. Could you please give me some advice regarding to troubleshooting this type of error? I really appreciate your great help!

#=========================#
Here is an example of my code:
queryList<-c("EBF1") # it was tested in the vignette
DefaultAssay(atac)<-"NarrowPeaks" # peaks called by MACS2
atac <- Footprint(
  object = atac,
  motif.name = queryList,
  genome = BSgenome.Hsapiens.UCSC.hg38,
  regions=NULL,
  assay=NULL,
  upstream=250,
  downstream=250,
  #compute.expected=TRUE,
  in.peaks=FALSE,
  verbose=TRUE
)
# Note: (1) Motif analysis has been done using this object. No errors were found. (2) Memory should be sufficient for running this job before getting the error message.

#==============#
# Error Messages:
Computing Tn5 insertion bias
Extracting reads in requested region
Extracting reads in requested region
Extracting reads in requested region
Error in InsertionBias.ChromatinAssay(object = object, genome = genome) : 
  Not all hexamers represented in input region
Calls: Footprint ... Footprint.ChromatinAssay -> InsertionBias -> InsertionBias.ChromatinAssay
In addition: There were 15 warnings (use warnings() to see them)
Execution halted
Warning message:
system call failed: Cannot allocate memory 
#==============#

> sessionInfo()
R version 4.1.1 (2021-08-10)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 16.04.7 LTS

Matrix products: default
BLAS:   /mnt/share/apps/tools/R/R-4.1.1/lib/R/lib/libRblas.so
LAPACK: /mnt/share/apps/tools/R/R-4.1.1/lib/R/lib/libRlapack.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C
 [9] LC_ADDRESS=C               LC_TELEPHONE=C
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods
[8] base

other attached packages:
 [1] motifmatchr_1.16.0                dplyr_1.0.7
 [3] plyr_1.8.6                        patchwork_1.1.1
 [5] scater_1.22.0                     scuttle_1.4.0
 [7] SingleCellExperiment_1.16.0       ggplot2_3.3.5
 [9] EnsDb.Hsapiens.v86_2.99.0         ensembldb_2.18.2
[11] AnnotationFilter_1.18.0           GenomicFeatures_1.46.1
[13] AnnotationDbi_1.56.2              Signac_1.4.0
[15] BSgenome.Hsapiens.UCSC.hg38_1.4.4 BSgenome_1.62.0
[17] rtracklayer_1.54.0                Biostrings_2.62.0
[19] XVector_0.34.0                    SeuratObject_4.0.3
[21] Seurat_4.0.5                      edgeR_3.36.0
[23] limma_3.50.0                      magrittr_2.0.1
[25] readr_2.1.0                       SummarizedExperiment_1.24.0
[27] Biobase_2.54.0                    GenomicRanges_1.46.1
[29] GenomeInfoDb_1.30.0               IRanges_2.28.0
[31] S4Vectors_0.32.3                  BiocGenerics_0.40.0
[33] MatrixGenerics_1.6.0              matrixStats_0.61.0
[35] Matrix_1.3-4

loaded via a namespace (and not attached):
  [1] utf8_1.2.2                  R.utils_2.11.0
  [3] reticulate_1.22             tidyselect_1.1.1
  [5] poweRlaw_0.70.6             RSQLite_2.2.8
  [7] htmlwidgets_1.5.4           grid_4.1.1
  [9] docopt_0.7.1                BiocParallel_1.28.1
 [11] Rtsne_0.15                  ScaledMatrix_1.2.0
 [13] munsell_0.5.0               codetools_0.2-18
 [15] ica_1.0-2                   future_1.23.0
 [17] miniUI_0.1.1.1              withr_2.4.3
 [19] colorspace_2.0-2            filelock_1.0.2
 [21] ROCR_1.0-11                 tensor_1.5
 [23] listenv_0.8.0               labeling_0.4.2
 [25] slam_0.1-49                 GenomeInfoDbData_1.2.7
 [27] polyclip_1.10-0             bit64_4.0.5
 [29] farver_2.1.0                parallelly_1.29.0
 [31] vctrs_0.3.8                 generics_0.1.1
 [33] BiocFileCache_2.2.0         lsa_0.73.2
 [35] ggseqlogo_0.1               R6_2.5.1
 [37] ggbeeswarm_0.6.0            rsvd_1.0.5
 [39] locfit_1.5-9.4              bitops_1.0-7
 [41] spatstat.utils_2.2-0        cachem_1.0.6
 [43] DelayedArray_0.20.0         assertthat_0.2.1
 [45] promises_1.2.0.1            BiocIO_1.4.0
 [47] scales_1.1.1                beeswarm_0.4.0
 [49] gtable_0.3.0                beachmat_2.10.0
 [51] globals_0.14.0              goftest_1.2-3
 [53] seqLogo_1.60.0              rlang_0.4.12
 [55] RcppRoll_0.3.0              splines_4.1.1
 [57] lazyeval_0.2.2              spatstat.geom_2.3-0
 [59] yaml_2.2.1                  reshape2_1.4.4
 [61] abind_1.4-5                 httpuv_1.6.3
 [63] tools_4.1.1                 ellipsis_0.3.2
 [65] spatstat.core_2.3-1         RColorBrewer_1.1-2
 [67] ggridges_0.5.3              Rcpp_1.0.7
 [69] sparseMatrixStats_1.6.0     progress_1.2.2
 [71] zlibbioc_1.40.0             purrr_0.3.4
 [73] RCurl_1.98-1.5              prettyunits_1.1.1
 [75] rpart_4.1-15                deldir_1.0-6
 [77] viridis_0.6.2               pbapply_1.5-0
 [79] cowplot_1.1.1               zoo_1.8-9
 [81] ggrepel_0.9.1               cluster_2.1.2
 [83] data.table_1.14.2           scattermore_0.7
 [85] lmtest_0.9-39               RANN_2.6.1
 [87] SnowballC_0.7.0             ProtGenerics_1.26.0
 [89] fitdistrplus_1.1-6          hms_1.1.1
 [91] mime_0.12                   xtable_1.8-4
 [93] XML_3.99-0.8                sparsesvd_0.2
 [95] gridExtra_2.3               compiler_4.1.1
 [97] biomaRt_2.50.1              tibble_3.1.6
 [99] KernSmooth_2.23-20          crayon_1.4.2
[101] R.oo_1.24.0                 htmltools_0.5.2
[103] mgcv_1.8-38                 later_1.3.0
[105] tzdb_0.2.0                  TFBSTools_1.32.0
[107] tidyr_1.1.4                 DBI_1.1.1
[109] tweenr_1.0.2                dbplyr_2.1.1
[111] MASS_7.3-54                 rappdirs_0.3.3
[113] R.methodsS3_1.8.1           parallel_4.1.1
[115] igraph_1.2.8                pkgconfig_2.0.3
[117] TFMPvalue_0.0.8             GenomicAlignments_1.30.0
[119] plotly_4.10.0               spatstat.sparse_2.0-0
[121] xml2_1.3.2                  annotate_1.72.0
[123] DirichletMultinomial_1.36.0 vipor_0.4.5
[125] stringr_1.4.0               digest_0.6.28
[127] pracma_2.3.3                sctransform_0.3.2
[129] RcppAnnoy_0.0.19            CNEr_1.30.0
[131] spatstat.data_2.1-0         leiden_0.3.9
[133] fastmatch_1.1-3             uwot_0.1.10
[135] DelayedMatrixStats_1.16.0   restfulr_0.0.13
[137] curl_4.3.2                  gtools_3.9.2
[139] shiny_1.7.1                 Rsamtools_2.10.0
[141] rjson_0.2.20                lifecycle_1.0.1
[143] nlme_3.1-153                jsonlite_1.7.2
[145] BiocNeighbors_1.12.0        viridisLite_0.4.0
[147] fansi_0.5.0                 pillar_1.6.4
[149] lattice_0.20-45             GO.db_3.14.0
[151] KEGGREST_1.34.0             fastmap_1.1.0
[153] httr_1.4.2                  survival_3.2-13
[155] glue_1.5.1                  qlcMatrix_0.9.7
[157] png_0.1-7                   bit_4.0.4
[159] ggforce_0.3.3               stringi_1.7.5
[161] blob_1.2.2                  BiocSingular_1.10.0
[163] caTools_1.18.2              memoise_2.0.0
[165] irlba_2.3.3                 future.apply_1.8.1

**Comment by timoast** [comment]
To compute the expected number of insertions, we first count the observed Tn5 insertion frequency at each DNA hexamer. To do this, by default we look only at the first chromosome. The error `Not all hexamers represented in input region` indicates that we couldn't find and instance of Tn5 inserting in every possible hexamer using the first chromosome, which can happen if your data is extremely sparse. You can change by running `InsertionBias()` and setting the `region` parameter. Alternatively, you can prevent the function from computing expected insertions by setting `compute.expected=FALSE`.

You seem to have another warning `system call failed: Cannot allocate memory` which indicates you might not have enough memory.

**Comment by timoast** [comment]
> The footprint function seems to be really memory- and time-consuming even when I used a small-scale subset of object, but with large fragment file. Maybe I should subset the fragment file (removing other cells not included in this test set) before running FootPrint?

The footprint function is memory intensive because it's constructing a cell x motif matrix which can end up quite large depending on the number of motifs. You can reduce the number of motifs used by setting `in.peaks=TRUE` which can substantially reduce the size of the matrix. Removing unused cells from the fragment file won't have any effect on memory usage.

**Comment by Jackieshen** [comment]
Hi, Tim,

Thanks for your useful advice, which I really appreciate. My case is kind of special. I did multiple-rounds of data integration, split/subset, and merge objects, therefore the fragments object stored on test set turns to be a list of fragment files, each pointing to unique cell barcodes. In fact, all the cell barcodes could be found in one fragment file and there is no need to enter iteration when performing MultiGetReadsInRegion.  I've re-built fragments object for this test set only containing one fragment filePath and it saved some time for the step of "Extracting reads in requested region". 

However, scanTabix() is still memory- and time-consuming, if I use large fragments object but query for small-scale test set. That's why I was wondering that whether subset of fragments file will help facilitate this step of calculation. Also, is there anyway to release the memory after performing GetReadsInRegion? My test set only contains very small cell number which should not require large memory after getting the reads from MultiGetReadsInRegion. But the memory-consumption didn't drop after completing this step, which leads to the memory issue for the sequential function of RunFootprint. I've tried to use gc() to release memory manually but it doesn't work well. Could you please give me some advice regarding to this issue?  Thank you very much!

**Comment by Jackieshen** [comment]
Hi, Tim,

I just finished test using re-constructed fragments object which only contains query cells. It speeds up the performance of InsertionBias() significantly and no longer requests large memory for computing. Thanks for your advice! My initial issue "Error in InsertionBias.ChromatinAssay(object = object, genome = genome)" has been solved by setting the region parameter for InsertionBias(). All your suggestion is really helpful, which I really appreciate. Thank you very much!
